lib/common/ipc_client.c

```diff
@@ -31,6 +31,672 @@
+    api->ipc = NULL;
+    api->mainloop_io = NULL;
+
+    if (api->free_on_disconnect) {
+        /* pcmk_free_ipc_api() has already been called, but did not free api
+         * or api->cmds because this function needed them. Do that now.
+         */
+        free_daemon_specific_data(api);
+        crm_trace("Freeing IPC API object after disconnect");
+        free(api);
+    }
+}
+
+/*!
+ * \brief Free the contents of an IPC API object
+ *
+ * \param[in] api  IPC API object to free
+ */
+void
+pcmk_free_ipc_api(pcmk_ipc_api_t *api)
+{
+    bool free_on_disconnect = false;
+
+    if (api == NULL) {
+        return;
+    }
+    crm_debug("Releasing %s IPC API", pcmk_ipc_name(api, true));
+
+    if (api->ipc != NULL) {
+        if (api->mainloop_io != NULL) {
+            /* We need to keep the api pointer itself around, because it is the
+             * user data for the IPC client destroy callback. That will be
+             * triggered by the pcmk_disconnect_ipc() call below, but it might
+             * happen later in the main loop (if still running).
+             *
+             * This flag tells the destroy callback to free the object. It can't
+             * do that unconditionally, because the application might call this
+             * function after a disconnect that happened by other means.
+             */
+            free_on_disconnect = api->free_on_disconnect = true;
+        }
+        pcmk_disconnect_ipc(api); // Frees api if free_on_disconnect is true
+    }
+    if (!free_on_disconnect) {
+        free_daemon_specific_data(api);
+        crm_trace("Freeing IPC API object");
+        free(api);
+    }
+}
+
+/*!
+ * \brief Get the IPC name used with an IPC API connection
+ *
+ * \param[in] api      IPC API connection
+ * \param[in] for_log  If true, return human-friendly name instead of IPC name
+ *
+ * \return IPC API's human-friendly or connection name, or if none is available,
+ *         "Pacemaker" if for_log is true and NULL if for_log is false
+ */
+const char *
+pcmk_ipc_name(pcmk_ipc_api_t *api, bool for_log)
+{
+    if (api == NULL) {
+        return for_log? "Pacemaker" : NULL;
+    }
+    switch (api->server) {
+        case pcmk_ipc_attrd:
+            return for_log? "attribute manager" : NULL /* T_ATTRD */;
+
+        case pcmk_ipc_based:
+            return for_log? "CIB manager" : NULL /* PCMK__SERVER_BASED_RW */;
+
+        case pcmk_ipc_controld:
+            return for_log? "controller" : NULL /* CRM_SYSTEM_CRMD */;
+
+        case pcmk_ipc_execd:
+            return for_log? "executor" : NULL /* CRM_SYSTEM_LRMD */;
+
+        case pcmk_ipc_fenced:
+            return for_log? "fencer" : NULL /* "stonith-ng" */;
+
+        case pcmk_ipc_pacemakerd:
+            return for_log? "launcher" : NULL /* CRM_SYSTEM_MCP */;
+
+        case pcmk_ipc_schedulerd:
+            return for_log? "scheduler" : NULL /* CRM_SYSTEM_PENGINE */;
+
+        default:
+            return for_log? "Pacemaker" : NULL;
+    }
```

Comment:
Should some of these be returning a string instead of NULL in the for_log == false case?

Reply:
Here, NULL means unimplemented. As we implement each daemon's IPC with this model, the corresponding string will be uncommented

Reply:
Right, I noticed that reviewing another patch in the series.  So that's fine.